File: nbi3gc-proxy/src/main/java/com/nsn/oss/nbi/AlarmInformationIteratorImpl.java
package com.nsn.oss.nbi;

import com.nsn.oss.nbi.common.Logger;
import org.omg.CosNotification.EventBatchHolder;
import org.omg.CosNotification.StructuredEvent;

import com.nsn.oss.nbi.corba.AlarmIRPSystem.AlarmInformationIteratorPOA;
import com.nsn.oss.nbi.corba.AlarmIRPSystem.NextAlarmInformations;
import com.nsn.oss.nbi.corba.ManagedGenericIRPSystem.InvalidParameter;
import com.nsn.oss.nbi.internal.corba.AlarmIRPSystem.AlarmInformationIterator;


public class AlarmInformationIteratorImpl extends AlarmInformationIteratorPOA {

    private static final Logger LOGGER = Logger.getLogger(AlarmInformationIteratorImpl.class);
    private AlarmInformationIterator reference;

    public AlarmInformationIteratorImpl(AlarmInformationIterator reference) {
        this.reference = reference;
    }

    @Override
    public boolean next_alarmInformations(short howManySource, EventBatchHolder alarmInformationsSource)
        throws NextAlarmInformations, InvalidParameter {
        LOGGER.info("Begin to getNext: " + howManySource);
        try {
            long beforeInvoke = System.currentTimeMillis();
            boolean result = reference.next_alarmInformations(howManySource, alarmInformationsSource);
            StructuredEvent[] structuredEventArraySource = alarmInformationsSource.value;
            //alarmInformationsSource.value = structuredEventArraySource;
            LOGGER.info("Complete to getNext, size: " + structuredEventArraySource.length + " result: " + result);
            long afterInvoke = System.currentTimeMillis();
            LOGGER.trace("time spent on next_alarmInformations is:" + (afterInvoke - beforeInvoke));
            return result;
        } catch (com.nsn.oss.nbi.internal.corba.AlarmIRPSystem.NextAlarmInformations e) {
            LOGGER.error("Fail to get next", e);
            throw new NextAlarmInformations("Fail to get next");
        } catch (com.nsn.oss.nbi.internal.corba.ManagedGenericIRPSystem.InvalidParameter e) {
            LOGGER.error("Fail to get next", e);
            throw new InvalidParameter(e.parameter);
        } catch (Exception e) {
            LOGGER.error("Fail to get next", e);
            return false;
        }
    }

    @Override
    public void destroy() {
        LOGGER.debug("destroy AlarmInformationIterator: " + reference);
        reference.destroy();
        //this._orb().shutdown(false );
    }

}

File: NextGenAct/src/main/scala/yang/iterator/AlarmInformationIteratorImpl.scala
package yang.iterator

import akka.actor.ActorRef
import akka.pattern.ask
import akka.util.Timeout
import com.nsn.oss.nbi.common.Logger
import com.nsn.oss.nbi.corba.AlarmIRPSystem.{AlarmInformationIteratorPOA, NextAlarmInformations}
import org.omg.CosNotification.{EventBatchHolder, StructuredEvent}
import yang.iterator.IteratorProtocol._

import scala.concurrent.Await
import scala.concurrent.duration._

/**
  * Created by y28yang on 2/16/2016.
  */
class AlarmInformationIteratorImpl(val iteratorActor: ActorRef, timeoutSec: Long) extends AlarmInformationIteratorPOA with TimeoutCheckable {
  val LOGGER = Logger.getLogger("AlarmInformationIteratorImpl")
  implicit val timeout = Timeout(timeoutSec seconds)

  //test for remote git
  override def destroy(): Unit = {
    iteratorActor ! RequestDestroyIterator
  }

  override def next_alarmInformations(howMany: Short, eventBatchHolder: EventBatchHolder): Boolean = {
    touch()
    val futureResult = iteratorActor ? request_next_date(howMany)
    try {
      val respond = Await.result(futureResult, timeout.duration).asInstanceOf[RespondNextDate[StructuredEvent]]
      eventBatchHolder.value = respond.structuredEvents.toArray
      respond.hasNext
    } catch {
      case e: Exception =>
        LOGGER.error("Fail to next alarm": Any, e: Throwable)
        throw new NextAlarmInformations(e.getMessage)

    }
  }
}
